subprojects/launcher/src/main/java/org/gradle/launcher/exec/RunActionRequirements.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.launcher.exec;
+
+import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.internal.hash.Hasher;
+
+public class RunActionRequirements extends AbstractToolingModelRequirements {
+    public RunActionRequirements(StartParameterInternal startParameter,
+                                 boolean runsTasks) {
+        super(startParameter, runsTasks);
+    }
+
+    @Override
+    public void appendKeyTo(Hasher hasher) {
+        // Identify the type of action
+        hasher.putByte((byte) 3);
```

Comment:
We have `Hasher.signature(Class)` to create a `HashCode` for the type, and then you can use `hasher.putHash()`. It is a little bit more costly, but it avoids a whole suite of problems with colliding IDs.

We could also move this thing to the execution engine, and mark these things as `@Input` properties, and let it handle caching on its own.